hazenlib/tasks/snr_map.py

```diff
@@ -1,11 +1,352 @@
+"""
+Map of local SNR across a flood phantom slice.
+
+Introduction
+============
+
+The SNR for each voxel in an image (of a flood phantom) is estimated as the SNR
+of a ROI centred on that voxel following the single image SNR method of McCann
+et al. [1]_. The SNR map can show variation in SNR caused by smoothing filters.
+It also highlights small regions of low signal which could be caused by micro-
+bubbles or foreign bodies in the phantom. These inhomogeneities can erroneously
+reduce SNR measurements made by other methods.
+
+
+Algorithm overview
+==================
+1. Apply boxcar smoothing to original image to create smooth image.
+2. Create noise image by subtracting smooth image from original image.
+3. Create image mask to remove background using e.g.
+     skimage.filters.threshold_minimum
+4. Calculate SNR using McCann's method and overlay ROIs on image.
+5. Estimate local noise as standard deviation of pixel values in ROI centred on
+    a pixel. Repeat for each pixel in the noise image.
+6. Plot the local noise as a heat map.
+
+
+.. [1] McCann, A. J., Workman, A., & McGrath, C. (2013). A quick and robust
+method for measurement of signal-to-noise ratio in MRI. Physics in Medicine
+& Biology, 58(11), 3775.
+"""
 from hazenlib.HazenTask import HazenTask
 
+import pathlib
+
+import pydicom
+import numpy as np
+import matplotlib.pyplot as plt
+import matplotlib.patches as patches
+import os
+
+import skimage.morphology
+from scipy import ndimage
+from skimage import filters
+
+from hazenlib.logger import logger
+# helper functions
+def sample_std(vals):
+    """
+    Return sample stdev using numpy.stdev(..., ddof=1).
+
+    Parameters
+    ----------
+    vals : array-like
+        Array-like object to calculate sample stdev.
+
+    Returns
+    -------
+    standard deviation : ndarray, same dtype as vals.
+    """
+    return np.std(vals, ddof=1)
+
```

Comment:
Why is this function outside the class?

Reply:
The function is used as an argument to `scipy.ndimage.filters.generic_filter` on L127, so it can't be a method (as the argument list would include `self`.

I try changing it to a `@staticmethod` within the class next week.

Reply:
ok, shall I then wait before merging?

Reply:
I'm happy either way. It should only take a few seconds to change (and a few minutes to test), but I need to do it from my home machine. I think it's a design decision that I don't have enough experience to know which way is 'best' and will follow your preference.

I should be able to correct it before the end of Tuesday at the latest. Let me know what you think.

Reply:
thanks @pcw24601 for letting me know. We are planning a new release on Thursday so I will keep this PR open until then. If you fix it on Tuesday then I'll merge it then and it will contain the fix, if not I will merge it Thursday as it is in the new release and your edit will be added to the next release at the end of February.

Reply:
I've removed the standard deviation function from outside the class. It was such a small function, and only used once, so I replaced it with a lambda function.